### ISSUE TITLE
improve diagnostics for possibly-unbound call-dunder with union

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#940]
When we have an implicit dunder call with a union type, where some types in the union implement the dunder and others do not, we currently report an error that the dunder methods are possibly unbound: https://play.ty.dev/e412c8da-cc72-47b5-af98-9db45800ff73

```py
class Context:
    def __enter__(self): ...
    def __exit__(self, *args): ...

class NotContext:
    pass

def _(x: Context | NotContext):
    # error: [invalid-context-manager] "Object of type `Context | NotContext` cannot be used with `with` because the methods `__enter__` and `__exit__` are possibly unbound"
    with x:
        pass
```

This is not very useful in identifying which element of the union type is the problem. It would be better if we would instead surface the underlying error directly (that `NotContext` does not implement `__enter__` or `__exit__`), and add a note that `NotContext` appears in the union type `Context | NotContext`.

This issue applies at least to sync and async context managers; I think it likely applies to some other implicit dunder calls as well.

**Comment by Y3K** [comment]
Hi, I'll try to fix this one.

**Comment by carljm** [comment]
This is the method-call variant of https://github.com/astral-sh/ty/issues/1656

**Comment by AlexWaygood** [comment]
> This is the method-call variant of [#1656](https://github.com/astral-sh/ty/issues/1656)

Sort-of? I think it might be more precise to say that both this issue and #1656 would probably require us to partially implement https://github.com/astral-sh/ty/issues/190?

**Comment by sharkdp** [comment]
If someone returns to working on this, see https://github.com/astral-sh/ruff/pull/20199 for a previous attempt to fix this.

**Comment by lerebear** [comment]
I've now proposed https://github.com/astral-sh/ruff/pull/24662 to solve this problem, although it only addresses the issue for context managers. If feedback on that use case is positive, then I will extend the same approach to the other use cases.

**Comment by lerebear** [comment]
> If feedback on that use case is positive, then I will extend the same approach to the other use cases.

https://github.com/astral-sh/ruff/pull/24676 addresses the remaining cases of this issue (for `__await__`, `__iter__` and `__mro_entries__`).

**Comment by lerebear** [comment]
I accidentally closed this prematurely; https://github.com/astral-sh/ruff/pull/24816 should actually address the last few instances of this issue (unary operators).
